pkg/operator/controller/externaldns/deployment.go

```diff
@@ -422,16 +417,18 @@ func externalDNSVolumesChanged(current, expected, updated *appsv1.Deployment) bo
 	currentVolumeMap := buildIndexedVolumeMap(current.Spec.Template.Spec.Volumes)
 	expectedVolumeMap := buildIndexedVolumeMap(expected.Spec.Template.Spec.Volumes)
 
+	// ignore defaultMode to prevent the fight with the defaulting
+	ignoreFieldsConfigMap := cmpopts.IgnoreFields(corev1.ConfigMapVolumeSource{}, "DefaultMode")
+	ignoreFieldsSecret := cmpopts.IgnoreFields(corev1.SecretVolumeSource{}, "DefaultMode")
+
 	// ensure all expected volumes are present,
 	// unsolicited ones are kept (e.g. kube api token)
 	for expName, expVol := range expectedVolumeMap {
 		if currVol, currExists := currentVolumeMap[expName]; !currExists {
 			updated.Spec.Template.Spec.Volumes = append(updated.Spec.Template.Spec.Volumes, expVol.Volume)
 			changed = true
 		} else {
-			// deepequal is fine here as we don't have more than 1 item
-			// neither in the secret nor in the configmap
-			if !reflect.DeepEqual(currVol.Volume, expVol.Volume) {
+			if !cmp.Equal(currVol.Volume, expVol.Volume, ignoreFieldsConfigMap, ignoreFieldsSecret) {
```

Comment:
Before we changed cluster-ingress-operator to hash the deployment, cluster-ingress-operator similarly used go-cmp to compare volumes—it might be worth a look to see whether there are any logic or unit tests worth adapting to external-dns-operator:

https://github.com/openshift/cluster-ingress-operator/blob/5a82b658c4952e8af5d4658d13e8802b93ded870/pkg/operator/controller/ingress/deployment.go#L359
https://github.com/openshift/cluster-ingress-operator/pull/182/commits/48c17ff3ab3dda4d259d5f26d196490c7ff89d12
https://github.com/openshift/cluster-ingress-operator/pull/182/commits/d32578a2d7e6628fdcb0487fdd941dbaef0c89fe
https://github.com/openshift/cluster-ingress-operator/pull/293/commits/5a82b658c4952e8af5d4658d13e8802b93ded870

For example, the ["if the volumes change ordering" test case](https://github.com/openshift/cluster-ingress-operator/pull/182/commits/48c17ff3ab3dda4d259d5f26d196490c7ff89d12#diff-c008ca2c033fa1a43f2d74d0af1febffdfbaf62e9fae623c90f38e4ecdf979edR333-R340) might be a good one to add to external-dns-operator.  

Cluster-ingress-operator's `cmpSecretVolumeSource` and `cmpConfigMapVolumeSource` also explicitly tested each field in the volume source types, which would reduce the likelihood of problems resulting from defaulting if upstream added new fields in these types.  Explicitly testing each field might not be worth the bother though.

Reply:
Ok, now I see where Steve was inspired from :)

> For example, the ["if the volumes change ordering" test case](https://github.com/openshift/cluster-ingress-operator/pull/182/commits/48c17ff3ab3dda4d259d5f26d196490c7ff89d12#diff-c008ca2c033fa1a43f2d74d0af1febffdfbaf62e9fae623c90f38e4ecdf979edR333-R340) might be a good one to add to external-dns-operator.

The order of the volumes doesn't matter in this comparison as we construct maps from the volume slices and compare by keys, added the test cases to confirm that though.

> Cluster-ingress-operator's cmpSecretVolumeSource and cmpConfigMapVolumeSource also explicitly tested each field in the volume source types, which would reduce the likelihood of problems resulting from defaulting if upstream added new fields in these types. Explicitly testing each field might not be worth the bother though.

To me, the real added value of `cmpSecretVolumeSource` and `cmpConfigMapVolumeSource` is the check of the `defaultMode` if it's set in the desired deployment. `EquateEmpty` options works well [even if specified](https://go.dev/play/p/59LQEP5eXW-) on the top level of the comparison chain, so I added it on the `Volume` level. I'll keep it ignored for the moment, let's revisit the comparison once we'll need to set desired value for it.